File: lib_view/src/main/java/com/xxf/view/utils/ResourcesUtil.java
package com.xxf.view.utils;


import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.Drawable;
import android.text.TextUtils;

import androidx.annotation.ColorInt;
import androidx.annotation.ColorRes;
import androidx.annotation.DrawableRes;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.core.content.ContextCompat;

import com.xxf.arch.XXF;
import com.xxf.arch.utils.EncryptUtils;
import com.xxf.arch.utils.FileUtils;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.UnsupportedEncodingException;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

/**
 * @Author: XGod  xuanyouwu@163.com  17611639080  https://github.com/NBXXF     https://blog.csdn.net/axuanqq  xuanyouwu@163.com  17611639080  https://github.com/NBXXF     https://blog.csdn.net/axuanqq
 * @Description 资源文件工具类
 */
public class ResourcesUtil {

    private static Context getContext() {
        return XXF.getApplication();
    }

    public static String getString(@StringRes int resId) {
        return getContext().getString(resId);
    }

    public static String getString(@StringRes int resId, Object... formatArgs) {
        return getContext().getString(resId, formatArgs);
    }

    public static Drawable getDrawable(@DrawableRes int resId) {
        return ContextCompat.getDrawable(getContext(), resId);
    }

    @ColorInt
    public static int getColor(@ColorRes int resId) {
        return ContextCompat.getColor(getContext(), resId);
    }


    public static int getIdByName(String name) {
        return getContext().getResources().getIdentifier(name, "id", getContext().getPackageName());
    }


    public static int getStringIdByName(String name) {
        return getContext().getResources().getIdentifier(name, "string", getContext().getPackageName());
    }


    public static int getColorIdByName(String name) {
        return getContext().getResources().getIdentifier(name, "color", getContext().getPackageName());
    }


    public static int getDimenIdByName(String name) {
        return getContext().getResources().getIdentifier(name, "dimen", getContext().getPackageName());
    }


    public static int getDrawableIdByName(String name) {
        return getContext().getResources().getIdentifier(name, "drawable", getContext().getPackageName());
    }


    public static int getMipmapIdByName(String name) {
        return getContext().getResources().getIdentifier(name, "mipmap", getContext().getPackageName());
    }


    public static int getLayoutIdByName(String name) {
        return getContext().getResources().getIdentifier(name, "layout", getContext().getPackageName());
    }


    public static int getStyleIdByName(String name) {
        return getContext().getResources().getIdentifier(name, "style", getContext().getPackageName());
    }


    public static int getAnimIdByName(String name) {
        return getContext().getResources().getIdentifier(name, "anim", getContext().getPackageName());
    }


    public static int getMenuIdByName(String name) {
        return getContext().getResources().getIdentifier(name, "menu", getContext().getPackageName());
    }


    public static boolean copyFileFromAssets(final String assetsFilePath, final String destFilePath) {
        boolean res = true;
        try {
            String[] assets = getContext().getAssets().list(assetsFilePath);
            if (assets != null && assets.length > 0) {
                for (String asset : assets) {
                    res &= copyFileFromAssets(assetsFilePath + "/" + asset, destFilePath + "/" + asset);
                }
            } else {
                res = FileUtils.writeFileFromIS(
                        new File(destFilePath),
                        getContext().getAssets().open(assetsFilePath), false
                );
            }
        } catch (IOException e) {
            e.printStackTrace();
            res = false;
        }
        return res;
    }


    public static String readAssets2String(final String assetsFilePath) {
        return readAssets2String(assetsFilePath, null);
    }


    public static String readAssets2String(final String assetsFilePath, @Nullable final String charsetName) {
        try {
            InputStream is = getContext().getAssets().open(assetsFilePath);
            byte[] bytes = FileUtils.getFileBytes(is);
            if (bytes == null) return "";
            if (TextUtils.isEmpty(charsetName)) {
                return new String(bytes);
            } else {
                try {
                    return new String(bytes, charsetName);
                } catch (UnsupportedEncodingException e) {
                    e.printStackTrace();
                    return "";
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
            return "";
        }
    }


    /**
     * 检查所有
     * 主module app R文件的内容是已经合并好的,但是有第三方依赖库里面的资源重复
     * 检查资源重复 第一步只检查字符串
     * 未来:id 颜色等等,请期待
     * 避免lint lint缺陷就是能忽略检查
     */
    public static void checkResources(List<Integer> ignoreIds) {
        Class<?> aClass = null;
        try {
            aClass = Class.forName(getContext().getPackageName() + ".R");
        } catch (ClassNotFoundException ex) {
            ex.printStackTrace();
        }
        if (aClass == null) {
            return;
        }
        List<Integer> stringResources = getStringResources(aClass);
        stringResources.removeAll(ignoreIds);
        /**
         * 字符串
         * key:string values:ids集合
         */
        Map<String, List<String>> stringRepeatMap = new HashMap<>();
        for (Integer id : stringResources) {
            String key = getString(id);
            List<String> idNames = stringRepeatMap.get(key);
            if (idNames == null) {
                idNames = new ArrayList<>();
            }
            idNames.add(getContext().getResources().getResourceEntryName(id));
            stringRepeatMap.put(key, idNames);
        }

        List<Integer> drawableResources = getDrawableResources(aClass);
        drawableResources.removeAll(ignoreIds);

        /**
         * 图片
         * key:drawable values:ids集合
         */
        Map<String, List<String>> drawableRepeatMap = new HashMap<>();
        for (Integer id : drawableResources) {
            /**
             * 会存在load 不出来的xml的图片
             */
            Bitmap bitmap = BitmapFactory.decodeResource(getContext().getResources(), id);
            if (bitmap != null) {
                bitmap = BitmapUtils.scale(bitmap, 32, 32);
                bitmap = BitmapUtils.grey(bitmap);
            } else {
                /**
                 * 加载xml中的drawable
                 */
                //bitmap = BitmapUtils.drawableToBitmap(getDrawable(id));
            }
            if (bitmap == null) {
                continue;
            }
            byte[] bytes = BitmapUtils.toByte(bitmap);
            BitmapUtils.recycle(bitmap);
            String key = null;
            if (bytes != null) {
                try {
                    key = EncryptUtils.encryptMD5ToString(bytes);
                } catch (Exception e) {
                    e.printStackTrace();
                }
                bytes = null;
            }
            if (TextUtils.isEmpty(key)) {
                continue;
            }
            List<String> idNames = drawableRepeatMap.get(key);
            if (idNames == null) {
                idNames = new ArrayList<>();
            }
            idNames.add(getContext().getResources().getResourceEntryName(id));
            drawableRepeatMap.put(key, idNames);
        }
        stringRepeatMap = convertRepeatMap(stringRepeatMap);
        StringBuffer repeatBuilder = new StringBuffer("String重复" + stringRepeatMap.size() + "条:");
        for (Map.Entry<String, List<String>> entry : stringRepeatMap.entrySet()) {
            repeatBuilder.append(entry.getValue());
            repeatBuilder.append(";");
        }


        drawableRepeatMap = convertRepeatMap(drawableRepeatMap);
        repeatBuilder.append("image相似或重复" + drawableRepeatMap.size() + "条:");
        for (Map.Entry<String, List<String>> entry : drawableRepeatMap.entrySet()) {
            repeatBuilder.append(entry.getValue());
            repeatBuilder.append(";");
        }

        if (!stringRepeatMap.isEmpty() || !drawableRepeatMap.isEmpty()) {
            throw new RepeatResourcesException(repeatBuilder.toString());
        }
    }

    private static Map<String, List<String>> convertRepeatMap(Map<String, List<String>> map) {
        if (map != null) {
            Iterator<Map.Entry<String, List<String>>> iterator = map.entrySet().iterator();
            while (iterator.hasNext()) {
                Map.Entry<String, List<String>> next = iterator.next();
                if (next.getValue() == null || next.getValue().size() <= 1) {
                    iterator.remove();
                }
            }
        }
        return map;
    }


    /**
     * 获取r文件的所有字符串资源ids
     *
     * @param rClazz
     * @return
     */
    public static List<Integer> getStringResources(Class... rClazz) {
        List<Integer> idList = new ArrayList<>();
        if (rClazz != null) {
            for (Class aClass : rClazz) {
                Class innerClazz[] = aClass.getDeclaredClasses();
                for (Class claszInner : innerClazz) {
                    /**
                     * find string.class
                     */
                    if (!TextUtils.equals(claszInner.getSimpleName(), "string")) {
                        continue;
                    }
                    Field[] fields = claszInner.getDeclaredFields();
                    for (Field field : fields) {
                        try {
                            int id = field.getInt(claszInner);
                            idList.add(id);
                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                    }
                }
            }
        }
        return idList;
    }

    /**
     * 获取指定R文件下面的所有drawable 资源ids
     *
     * @param rClazz
     * @return
     */
    public static List<Integer> getDrawableResources(Class... rClazz) {
        List<Integer> idList = new ArrayList<>();
        if (rClazz != null) {
            for (Class aClass : rClazz) {
                Class innerClazz[] = aClass.getDeclaredClasses();
                for (Class claszInner : innerClazz) {
                    if (TextUtils.equals(claszInner.getSimpleName(), "drawable") || TextUtils.equals(claszInner.getSimpleName(), "mipmap")) {
                        Field[] fields = claszInner.getDeclaredFields();
                        for (Field field : fields) {
                            try {
                                int id = field.getInt(claszInner);
                                idList.add(id);
                            } catch (Exception e) {
                                e.printStackTrace();
                            }
                        }
                    }
                }
            }
        }
        return idList;
    }

    public static class RepeatResourcesException extends RuntimeException {
        public RepeatResourcesException(String message) {
            super(message);
        }
    }
}

File: lib_utils/src/main/java/com/xxf/utils/ResourcesUtil.kt
package com.xxf.utils

import android.content.Context
import android.graphics.BitmapFactory
import android.graphics.drawable.Drawable
import android.text.TextUtils
import androidx.annotation.ColorInt
import androidx.annotation.ColorRes
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.core.content.ContextCompat
import com.xxf.application.applicationContext
import com.xxf.utils.BitmapUtils.grey
import com.xxf.utils.BitmapUtils.recycle
import com.xxf.utils.BitmapUtils.scale
import com.xxf.utils.BitmapUtils.toByte
import java.io.File
import java.io.IOException
import java.io.UnsupportedEncodingException
import java.nio.charset.Charset
import java.util.*

/**
 * @Author: XGod  xuanyouwu@163.com  17611639080  https://github.com/NBXXF     https://blog.csdn.net/axuanqq  xuanyouwu@163.com  17611639080  https://github.com/NBXXF     https://blog.csdn.net/axuanqq
 * @Description 资源文件工具类
 */
object ResourcesUtil {

    /**
     * @param context设计为可选参数 主要是VectorEnabledTintResources 和 TintResources 的影响
     */
    fun getString(@StringRes resId: Int,context: Context= applicationContext): String {
        return context.getString(resId)
    }

    /**
     * @param context设计为可选参数 主要是VectorEnabledTintResources 和 TintResources 的影响
     */
    fun getString(@StringRes resId: Int, vararg formatArgs: Any?,context: Context=applicationContext): String {
        return context.getString(resId, *formatArgs)
    }

    /**
     * @param context设计为可选参数 主要是VectorEnabledTintResources 和 TintResources 的影响
     */
    fun getDrawable(@DrawableRes resId: Int,context: Context=applicationContext): Drawable? {
        return ContextCompat.getDrawable(context!!, resId)
    }

    /**
     * @param context设计为可选参数 主要是VectorEnabledTintResources 和 TintResources 的影响
     */
    @ColorInt
    fun getColor(@ColorRes resId: Int,context: Context=applicationContext): Int {
        return ContextCompat.getColor(context!!, resId)
    }

    fun getIdByName(context: Context, name: String?): Int {
        return context.resources.getIdentifier(name, "id", context.packageName)
    }

    fun getStringIdByName(context: Context, name: String?): Int {
        return context.resources.getIdentifier(name, "string", context.packageName)
    }

    fun getColorIdByName(context: Context, name: String?): Int {
        return context.resources.getIdentifier(name, "color", context.packageName)
    }

    fun getDimenIdByName(context: Context, name: String?): Int {
        return context.resources.getIdentifier(name, "dimen", context.packageName)
    }

    fun getDrawableIdByName(context: Context, name: String?): Int {
        return context.resources.getIdentifier(name, "drawable", context.packageName)
    }

    fun getMipmapIdByName(context: Context, name: String?): Int {
        return context.resources.getIdentifier(name, "mipmap", context.packageName)
    }

    fun getLayoutIdByName(context: Context, name: String?): Int {
        return context.resources.getIdentifier(name, "layout", context.packageName)
    }

    fun getStyleIdByName(context: Context, name: String?): Int {
        return context.resources.getIdentifier(name, "style", context.packageName)
    }

    fun getAnimIdByName(context: Context, name: String?): Int {
        return context.resources.getIdentifier(name, "anim", context.packageName)
    }

    fun getMenuIdByName(context: Context, name: String?): Int {
        return context.resources.getIdentifier(name, "menu", context.packageName)
    }

    fun copyFileFromAssets(assetsFilePath: String, destFilePath: String): Boolean {
        var res = true
        try {
            val assets = applicationContext.assets.list(assetsFilePath)
            if (assets != null && assets.size > 0) {
                for (asset in assets) {
                    res = res and copyFileFromAssets("$assetsFilePath/$asset", "$destFilePath/$asset")
                }
            } else {
                res = FileUtils.writeFileFromIS(
                        File(destFilePath),
                     applicationContext.assets.open(assetsFilePath), false
                )
            }
        } catch (e: IOException) {
            e.printStackTrace()
            res = false
        }
        return res
    }

    @JvmOverloads
    fun readAssets2String(assetsFilePath: String?, charsetName: String? = null): String {
        return try {
            val `is` = applicationContext.assets.open(assetsFilePath!!)
            val bytes = FileUtils.getFileBytes(`is`) ?: return ""
            if (TextUtils.isEmpty(charsetName)) {
                String(bytes)
            } else {
                try {
                    String(bytes, Charset.forName(charsetName))
                } catch (e: UnsupportedEncodingException) {
                    e.printStackTrace()
                    ""
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
            ""
        }
    }

    /**
     * 检查所有
     * 主module app R文件的内容是已经合并好的,但是有第三方依赖库里面的资源重复
     * 检查资源重复 第一步只检查字符串
     * 未来:id 颜色等等,请期待
     * 避免lint lint缺陷就是能忽略检查
     */
    fun checkResources( ignoreIds: List<Int>?) {
        var aClass: Class<*>? = null
        try {
            aClass = Class.forName(applicationContext.packageName + ".R")
        } catch (ex: ClassNotFoundException) {
            ex.printStackTrace()
        }
        if (aClass == null) {
            return
        }
        val stringResources = getStringResources(aClass)
        stringResources.removeAll(ignoreIds!!)
        /**
         * 字符串
         * key:string values:ids集合
         */
        var stringRepeatMap: MutableMap<String?, MutableList<String?>?>? = HashMap()
        for (id in stringResources) {
            val key = getString(id)
            var idNames = stringRepeatMap!![key]
            if (idNames == null) {
                idNames = ArrayList()
            }
            idNames.add(applicationContext.resources.getResourceEntryName(id))
            stringRepeatMap[key] = idNames
        }
        val drawableResources = getDrawableResources(aClass)
        drawableResources.removeAll(ignoreIds)
        /**
         * 图片
         * key:drawable values:ids集合
         */
        var drawableRepeatMap: MutableMap<String?, MutableList<String?>?>? = HashMap()
        for (id in drawableResources) {
            /**
             * 会存在load 不出来的xml的图片
             */
            var bitmap = BitmapFactory.decodeResource(applicationContext.resources, id)
            if (bitmap != null) {
                bitmap = scale(bitmap, 32, 32)
                bitmap = grey(bitmap)
            } else {
                /**
                 * 加载xml中的drawable
                 */
                //bitmap = BitmapUtils.drawableToBitmap(getDrawable(id));
            }
            if (bitmap == null) {
                continue
            }
            var bytes = toByte(bitmap)
            recycle(bitmap)
            var key: String? = null
            if (bytes != null) {
                try {
                    key = EncryptUtils.encryptMD5ToString(bytes)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
                bytes = null
            }
            if (TextUtils.isEmpty(key)) {
                continue
            }
            var idNames = drawableRepeatMap!![key]
            if (idNames == null) {
                idNames = ArrayList()
            }
            idNames.add(applicationContext.resources.getResourceEntryName(id))
            drawableRepeatMap[key] = idNames
        }
        stringRepeatMap = convertRepeatMap(stringRepeatMap)
        val repeatBuilder = StringBuffer("String重复" + stringRepeatMap!!.size + "条:")
        for ((_, value) in stringRepeatMap) {
            repeatBuilder.append(value)
            repeatBuilder.append(";")
        }
        drawableRepeatMap = convertRepeatMap(drawableRepeatMap)
        repeatBuilder.append("image相似或重复" + drawableRepeatMap!!.size + "条:")
        for ((_, value) in drawableRepeatMap) {
            repeatBuilder.append(value)
            repeatBuilder.append(";")
        }
        if (!stringRepeatMap.isEmpty() || !drawableRepeatMap.isEmpty()) {
            throw RepeatResourcesException(repeatBuilder.toString())
        }
    }

    private fun convertRepeatMap(map: MutableMap<String?, MutableList<String?>?>?): MutableMap<String?, MutableList<String?>?>? {
        if (map != null) {
            val iterator: MutableIterator<Map.Entry<String?, List<String?>?>> = map.entries.iterator()
            while (iterator.hasNext()) {
                val next = iterator.next()
                if (next.value == null || next.value!!.size <= 1) {
                    iterator.remove()
                }
            }
        }
        return map
    }

    /**
     * 获取r文件的所有字符串资源ids
     *
     * @param rClazz
     * @return
     */
    @JvmStatic
    fun getStringResources(vararg rClazz: Class<*>): MutableList<Int> {
        val idList: MutableList<Int> = ArrayList()
        if (rClazz != null) {
            for (aClass in rClazz) {
                val innerClazz = aClass.declaredClasses
                for (claszInner in innerClazz) {
                    /**
                     * find string.class
                     */
                    if (!TextUtils.equals(claszInner.simpleName, "string")) {
                        continue
                    }
                    val fields = claszInner.declaredFields
                    for (field in fields) {
                        try {
                            val id = field.getInt(claszInner)
                            idList.add(id)
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }
                    }
                }
            }
        }
        return idList
    }

    /**
     * 获取指定R文件下面的所有drawable 资源ids
     *
     * @param rClazz
     * @return
     */
    @JvmStatic
    fun getDrawableResources(vararg rClazz: Class<*>): MutableList<Int> {
        val idList: MutableList<Int> = ArrayList()
        if (rClazz != null) {
            for (aClass in rClazz) {
                val innerClazz = aClass.declaredClasses
                for (claszInner in innerClazz) {
                    if (TextUtils.equals(claszInner.simpleName, "drawable") || TextUtils.equals(claszInner.simpleName, "mipmap")) {
                        val fields = claszInner.declaredFields
                        for (field in fields) {
                            try {
                                val id = field.getInt(claszInner)
                                idList.add(id)
                            } catch (e: Exception) {
                                e.printStackTrace()
                            }
                        }
                    }
                }
            }
        }
        return idList
    }

    class RepeatResourcesException(message: String?) : RuntimeException(message)
}